### PR TITLE
build: switch from required to imports in custom lib modules

### PR DESCRIPTION
**Pull request by eiri**
## PR summary

This is a build refactoring work that makes sure we are using esm style imports in all of our custom modules. 

## PR Checklist

Please make sure that your PR fulfills the following requirements:

- [x] The commit message follows the
[Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type  

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [x] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

## What is the current behavior?

We have a mix of `import` and `required` statements. 

## What is the new behavior?

We still have `required` in generated modules, but our custom modules are now have consistent imports.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -110,44 +119,19 @@ module.exports = {
...
     {
-      'files': ['test/**/*.test.js'],
+      'files': ['test/**/*.test.js', 'test/**/*.test.mjs'],
       'rules': {
+        'import/namespace': 'off',
         'jest/expect-expect': 'off',
         'jest/no-conditional-expect': 'off',
         'jest/no-done-callback': 'off',
-        'jest/no-standalone-expect': 'off',
-        'jest/no-try-expect': 'off',
```
Comment (mojito317) [inline .eslintrc.js:119]: Are these rules not enforced anymore?

File: .eslintrc.js [hunk 2]
```diff
@@ -110,44 +119,19 @@ module.exports = {
...
       },
     },
     {
-      'files': ['test/**/cloudant.v1.test.js'],
+      'files': ['test/**/cloudant.*.test.js', 'test/**/auth-helper.js'],
       'rules': {
+        '@typescript-eslint/no-require-imports': 'off',
         'prettier/prettier': 'off',
-        'object-shorthand': 'off',
```
Comment (mojito317) [inline .eslintrc.js:126]: Is this rule not enforced anymore?

File: .eslintrc.js [hunk 3]
```diff
@@ -110,44 +119,19 @@ module.exports = {
...
-        'dot-notation': 'off',
-      },
-    },
-    {
-      'files': [
-        'test/examples/src/js/CreateDbAndDoc.js',
-        'test/examples/src/ts/CreateDbAndDoc.ts',
-      ],
-      'rules': {
-        'prefer-template': 'off',
```
Comment (mojito317) [inline .eslintrc.js:150]: Could you turn file extensions of `.js` into `.mjs` instead of removing this block?

File: tsconfig.json [hunk 4]
```diff
@@ -21,6 +21,7 @@
     /* Module Resolution Options */
     "moduleResolution": "node16",          /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
     "types": ["node"],                     /* Type declaration files to be included in compilation. */
+    "resolveJsonModule": true,
```
Comment (mojito317) [inline tsconfig.json:24]:
```suggestion
    "resolveJsonModule": true,             /* Allows importing modules with a .json extension. Helpful for version check.*/
```

**Review by mojito317 — COMMENTED** [review]
A couple of questions and change suggestions.

File: test/integration/readme.integration.test.mjs [hunk 5]
```diff
@@ -13,15 +13,13 @@
...
-const http = require('http');
-const fs = require('fs');
-const sinon = require('sinon');
-const {
-  getInfoFromExistingDatabase,
-} = require('../examples/src/js/GetInfoFromExistingDatabase.js');
-const { createDbAndDoc } = require('../examples/src/js/CreateDbAndDoc.js');
-const { updateDoc } = require('../examples/src/js/UpdateDoc.js');
-const { deleteDoc } = require('../examples/src/js/DeleteDoc.js');
+import { request as _request } from 'node:http';
```
[the remark this answers is not in the document]
Comment (eiri) [reply test/integration/readme.integration.test.mjs]: Done in 753c58e9677cbec1007f9bbefcb0d36f44f8a5e2

File: test/integration/timeout.integration.test.mjs [hunk 6]
```diff
@@ -14,16 +14,11 @@
  * limitations under the License.
  */
 
-const assert = require('assert');
-const sinon = require('sinon');
+import { ok, equal } from 'node:assert';
+import { spy, stub, assert as _assert, match } from 'sinon';
```
[the remark this answers is not in the document]
Comment (eiri) [reply test/integration/timeout.integration.test.mjs]: Done in 753c58e9677cbec1007f9bbefcb0d36f44f8a5e2

File: test/unit/cloudantBaseService.test.mjs [hunk 7]
```diff
@@ -14,18 +14,18 @@
...
  */
 
-const assert = require('assert');
-const http = require('http');
-const https = require('https');
-const {
+import { ok, strictEqual, notDeepStrictEqual } from 'node:assert';
+import { Agent } from 'node:http';
+import { Agent as _Agent } from 'node:https';
+import {
```
[the remark this answers is not in the document]
Comment (eiri) [reply test/unit/cloudantBaseService.test.mjs]: Done in 753c58e9677cbec1007f9bbefcb0d36f44f8a5e2

File: test/unit/getAuthenticatorFromEnvCloudantExtension.test.mjs [hunk 8]
```diff
@@ -13,11 +13,11 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-const assert = require('assert');
-const { BasicAuthenticator } = require('ibm-cloud-sdk-core');
-const getAuthenticatorFromEnvCloudantExtension =
-  require('../../lib/getAuthenticatorFromEnvCloudantExtension.ts').default;
-const { CouchdbSessionAuthenticator } = require('../../index.ts');
+import { ok, strictEqual } from 'node:assert';
```
[the remark this answers is not in the document]
Comment (eiri) [reply test/unit/getAuthenticatorFromEnvCloudantExtension.test.mjs]: Done in 753c58e9677cbec1007f9bbefcb0d36f44f8a5e2

File: test/unit/sessionTokenManager.test.mjs [hunk 9]
```diff
@@ -14,10 +14,10 @@
  * limitations under the License.
  */
 
-const assert = require('assert');
-const sinon = require('sinon');
-const { CookieJar } = require('tough-cookie');
-const { SessionTokenManager } = require('../../auth/sessionTokenManager.ts');
+import { strictEqual, ok, deepStrictEqual } from 'node:assert';
```
[the remark this answers is not in the document]
Comment (eiri) [reply test/unit/sessionTokenManager.test.mjs]: Done in 753c58e9677cbec1007f9bbefcb0d36f44f8a5e2

File: .eslintrc.js [hunk 10]
```diff
@@ -56,6 +63,7 @@ module.exports = {
       'plugins': ['node', 'prettier'],
```
[the remark this answers is not in the document]
Comment (eiri) [reply .eslintrc.js:63]: Not really. We only have `mjs` in tests and we have test specific rules section below

File: .eslintrc.js (hunk 1 above)
Comment (eiri) [reply .eslintrc.js:119]: That's other way around. These rules been off for some reason, but it is not necessary as linter passes without those overrides, so here we are enforcing them now.

File: .eslintrc.js (hunk 2 above)
Comment (eiri) [reply .eslintrc.js:126]: Same as above, this removes 'off', rule.

File: .eslintrc.js (hunk 3 above)
Comment (eiri) [reply .eslintrc.js:150]: As you can see they all are 'off' overrides and linter is passing without them, so this block been kind of meaningless.

File: tsconfig.json (hunk 4 above)
Comment (eiri) [reply tsconfig.json]: Done in 753c58e9677cbec1007f9bbefcb0d36f44f8a5e2

File: test/unit/couchdbSessionAuthenticator.test.mjs [hunk 11]
```diff
@@ -13,20 +13,20 @@
...
  */
-const assert = require('assert');
-const { promisify } = require('util');
-const sinon = require('sinon');
-const { CookieJar } = require('tough-cookie');
-const { CouchdbSessionAuthenticator } = require('../../index.ts');
-const { SessionTokenManager } = require('../../auth/sessionTokenManager.ts');
+import { strictEqual, ok } from 'node:assert';
+import { promisify } from 'node:util';
+import { stub, fake, replace, restore, useFakeTimers } from 'sinon';
```
[the remark this answers is not in the document]
Comment (eiri) [reply test/unit/couchdbSessionAuthenticator.test.mjs]: Done in 753c58e9677cbec1007f9bbefcb0d36f44f8a5e2

File: test/integration/timeout.integration.test.mjs [hunk 12]
```diff
@@ -123,7 +118,7 @@ describe('Default timeout config tests', () => {
       assert.ok(spyAuth.calledOnce);
       assert.ok(getTokenStubFn.calledOnce);
       // authenticate is called with default timeout
-      sinon.assert.calledWith(spyAuth, sinon.match.has('timeout', expValue));
+      sinon.assert.calledWith(spyAuth, match.has('timeout', expValue));
```
[the remark this answers is not in the document]
Comment (eiri) [reply test/integration/timeout.integration.test.mjs]: Guess not, given the test doesn't care. Anyway, fixed in 3e19f54a

File: test/unit/sessionTokenManager.test.mjs [hunk 13]
```diff
@@ -78,7 +78,7 @@ describe('SessionTokenManager tests', () => {
       );
       sendRequestStubFn.returnsArg(0);
       const parameters = manager.requestToken();
-      assert.deepStrictEqual(parameters, expectedParameters);
+      assert.strictEqual(parameters, expectedParameters);
```
[the remark this answers is not in the document]
Comment (eiri) [reply test/unit/sessionTokenManager.test.mjs]: My guess is that just strictEqual is passing the same, but fixed in 3e19f54a

File: .eslintrc.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (eiri) [reply .eslintrc.js:63]: Well, we can't change generated files that are still use `required`, so they are not mjs. Our manual lib files are `.ts`. Now, for files in `test/examples/src/features/js` I probably just missed the change the extension, they are hidden pretty deep 😉

**Review by mojito317 — APPROVED** [review]

**Comment by eiri** [comment]
@mojito317  it turned out jest was ignoring mjs tests and once I configured it not to it became really hard to make mix of cjs and esm import styles to work together in tests. So this PR superseded by #1629 that uses `imports` in lib and consistent `required` in tests. Sorry about the wasted review.

**Pull request closed without merging**